### DISCUSSION TITLE
PHP constraint ignored

### DISCUSSION THREAD

**Discussion by RolphH** [renovatebot/renovate#19555] [Request Help]
Hi all,

I noticed that (not sure if it is recently), Renovate ignores the contraints related to the `php` package specified in the `composer.json`. We are not using the `config.platform.php`, but rather use the `require.php` setting. For example (part of) composer.json:

```
{ "require": {
        "php": "~7.2.0", 
        "ext-json": "*" ,
        "symfony/monolog-bundle": "^3.1.0",
        "sensio/framework-extra-bundle": "^5.0.0"
}
```

Our Renovate configuration is like this:
```
{
  "rangeStrategy": "bump",
  "platform": "bitbucket-server",
  "endpoint": "https://********",
  "automerge": true,
  "lockFileMaintenance": { "enabled": true },
  "packageRules": [
  {
    "matchUpdateTypes": ["major"],
    "enabled": "false"
  },
  {
    "matchPackagePatterns": [
      "*"
    ],
    "matchUpdateTypes": [
      "patch"
    ],
    "groupName": "all patch updates",
    "groupSlug": "all-patch"
  },
  {
    "matchPackagePatterns": [
      "*"
    ],
    "matchUpdateTypes": [
      "minor"
    ],
    "groupName": "all minor updates",
    "groupSlug": "all-minor"
  }
  ],
  "repositories": [
    "repo1",
    "repo2"
  ]
}
```

When Renovate runs, for the minor updates, it also bumps the `php` version from `~7.4.33`:
```
{ "require": {
        "php": "~7.4.33",
        "ext-json": "*",
        "symfony/monolog-bundle": "^3.8.0",
        "sensio/framework-extra-bundle": "^5.6.1"
}
```
  I didn't expect this (as it `7.4.33` is not within the `~7.2.0` range). How can we prevent this from happening? All other packages are simply updated within the specified ranges.

**Comment by rarkins** [comment]
Renovate's purpose is to upgrade any dependencies if finds, and that's its default behavior. You can disable upgrades for whatever scenario you want using packageRules

**Comment by RolphH** [comment]
But it should follow the constraints right (eg: filter out those upgrades falling out of the range specified). And in this case `7.4.33` falls outside `~7.2.0` (eg: >= 7.2.0  < 7.3.0).
Renovate follows the constraints for all other packages, but seems not to do so for `php`

**Reply by rarkins** [reply]
If it "followed" constraints as you describe it then you'd never get a PR upgrading you beyond php 7.2?

**Reply by RolphH** [reply]
Correct, that was what I was expecting, but apparently it is not. I have now excluded php from minor updates. I guess you are correct that Renovate doesn't follow the constraints (which usually is fine I guess, but there are cases you might want to stick to a particular minor version (our use case: we run php 7.2.* on our servers as this is Ubuntu 18.04 LTS default. Bumping to 7.4 means we need to update to non-standard php version on Ubuntu 18.04 LTS, and as such we want to exclude the minor bump).

**Reply by rarkins** [reply]
I understand completely there are times you need to hold back versions (we do so in this repo too). It's just that Renovate's default behavior is to update everything it can, and then it's up to you to either ignore those, or suppress them (e.g. by setting dependencyDashboardApproval=true)
